### ISSUE TITLE
Release drafter selects wrong release as last release

### ISSUE THREAD

**Issue by TTvanWillegen** [release-drafter/release-drafter#913]
The release drafter thinks that V4.1.2 was the last release, however, our last release was V9.0.0. 
![image](https://user-images.githubusercontent.com/15815200/127441863-d1fb3024-29f4-49e3-aa90-23e15bc7ed51.png)
![image](https://user-images.githubusercontent.com/15815200/127441949-605c66e8-3728-4e6d-94b7-2619bb653524.png)

It has been doing this ever since drafting the first release with the release-drafter; back then the next version would indeed have been V5.0.0. Any later release has been manually adapted and released.

What could cause this / how can I fix this?

**Comment by neb-b** [comment]
@TTvanWillegen did you ever figure this out? It's been happening with us too. It keeps drafting an older version with all commits since that last version, even though there have been many releases since then

**Comment by neb-b** [comment]
I just figured out our issue. We had previous releases that were named by the date they were on ex `v2022.04.28`, then switched to semver with release-drafter. After deleting all of the old releases/tags its seems to be working correctly

**Comment by TTvanWillegen** [comment]
Hmm, all of our releases followed semver. Some earlier releases were released manually though, might be the problem there.

**Comment by oleg-nenashev** [comment]
So this is still actual, and I believe it happens because of this code: https://github.com/release-drafter/release-drafter/blob/ff929b5ceb21bf2646a216e916f9a8bb507d48a3/lib/releases.js#L8-L24

By default it delegates code to https://www.npmjs.com/package/compare-versions , after truncating the prefix that might have changed between the configurations. We use a 4-digit format like `3.0.0-beta-14` and eventually the library went wrong and seems to be doing just a string comparison of the suffix and hence resulting in "3.0.0-beta-9" being the most recent release.

Ideally the method should try to parse the version using the `version-format` value,  and then by comparing the parsed semver as a first step. Does it sound reasonable @jetersen ?

**Comment by 0rax** [comment]
I am currently facing related issue, and I feel like this might be related to the portion of code you sent @oleg-nenashev.

We have a monorepo that includes all of our backend components (which are tagged as `v1.x.x`) and our mobile app (which has separate tags, namely `app/v1.x.x`).
It seems that every time the app is released, any subsequent call to release-drafter will use this latest app version instead of just looking for releases with a valid prefix (here `^v` instead of `^app/v`).

This means seems to mean that, if there is ANY release that doesn't match the specified `version-prefix`, they will always be used first if they are more recent. I feel like using the publish date should be a fallback only if there isn't any valid tag found or at least there should be a way to enforce only tags matching the specified prefix / a specified format.

**Comment by oleg-nenashev** [comment]
I think it's definitely something we should change for the default behavior at least. I was about doing a patch but have never got to it

**Comment by lmagyar** [comment]
FYI: We fixed this issue in our fork, we use GitHub's latest release as possible oldest release when sorting releases to find last release. I won't create a PR here, but feel free to use this as you want: https://github.com/homeassistant-apps/action-release-drafter/pull/2

**Comment by AlexSkrypnyk** [comment]
## Analysis

Looking at the sortReleases function in releases.js:
```
  const sortReleases = (releases, tagPrefix) => {
    const tagPrefixRexExp = new RegExp(`^${regexEscape(tagPrefix)}`)
    return releases.sort((r1, r2) => {
      try {
        return compareVersions(
          r1.tag_name.replace(tagPrefixRexExp, ''),
          r2.tag_name.replace(tagPrefixRexExp, '')
        )
      } catch {
        return new Date(r1.created_at) - new Date(r2.created_at)
      }
    })
  }
```
  The compare-versions library compares versions numerically. When you have:
  - CalVer: 25.1.0 (meaning year 2025, month 1, patch 0)
  - SemVer: 1.2.3 (meaning major 1, minor 2, patch 3)

  The library compares them as numbers: 25.1.0 > 1.2.3 because 25 > 1.

  So the CalVer releases (like 25.1.0) will always be considered "higher" than your SemVer releases (like 
  1.2.3), regardless of which was actually released more recently.

  The lastRelease is determined by:
  const lastRelease = sortedSelectedReleases[sortedSelectedReleases.length - 1]

  This picks the last element after sorting, which will be the one with the "highest" version number - the
  CalVer release.

## Workarounds

1. Remove all CalVer releases
2. Get latest SemVer release outside of the action and pass it to the action as `version`

## Fixes at the library level

  **Idea 1: Add a version-pattern regex filter**

  Add a config option like version-pattern: '^[0-9]{1,2}\.[0-9]+\.[0-9]+$' that filters which releases are
  considered when determining lastRelease.

  This is more flexible than Idea 1 - users define their own pattern to match valid releases for their
  versioning scheme.

  **Idea 2: Add exclude-tags / include-tags pattern matching**

  Similar to existing exclude-labels, add:
  - exclude-tags: ['^2[0-9]\\.'] - exclude tags matching these patterns
  - include-tags: ['^[0-1]?[0-9]\\.'] - only include tags matching these patterns

  This would filter releases before the sortReleases function runs.

  **Idea 3: Change sorting to prefer creation date when versions are incompatible**

  Modify sortReleases to detect when comparing versions from different schemes (e.g., one has major >= 20,
  other has major < 20) and fall back to created_at comparison for those pairs.

  This is a heuristic approach - less explicit but would "just work" for mixed CalVer/SemVer repos.

  **Idea 4: Add latest-release-tag explicit override**

  Add a config option to explicitly specify which release to use as the base:
  latest-release-tag: '1.2.3'

  This bypasses the automatic detection entirely. Simple but requires manual updates.

**Comment by lmagyar** [comment]
> Idea 4: Add latest-release-tag explicit override

Idea 4+: Take GitHub's latest release, sort all the prereleases after that with the current code. It's backward compatible and solves this issue. See: https://github.com/homeassistant-apps/action-release-drafter/commit/c581fd250b8607c7dafece1caa68be1cfae765bd

**Comment by cchanche** [comment]
Reading the entier issue, I have a hard time parsing the exact scenarios in which the comparaison function would not produce the desired output.

I believe showcasing some of them may not be that hard to do in a small test. Such a (failing) test could then be the foundation for iterating on a PR for this.

As for a potential solution, I would love if we avoid adding a new input about tags / version patterns. I think those that exist are already quite confusing !

What I understand, is that the versioning scheme varies from one project to the other, sometimes using schemes different from semver (ex: calver) - which the current code seems not to have anticipated.

What about using a new `release-sorting-strategy` input, with the default value being `compare-versions` ?

This, for one, is non-breaking, and we could seize the opportunity to document properly how things work. (no need to look into the code)

I think this also encourages contributors to implement their own strategy right-off. From the discussion here, I can already see how `release-sorting-strategy: cal-ver` and `release-sorting-strategy: release-date` may be a thing.

**Comment by lmagyar** [comment]
> Reading the entier issue, I have a hard time parsing the exact scenarios in which the comparaison function would not produce the desired output.

Scenario:
- Developer uses semver, latest release is v1.1.1
- Then changes his mind, uses calver, latest release is eg. v2025.12.0.
- Then changes his mind again, uses semver again, the next release is v1.2.3.

This v1.2.3 release will never ever be detected as last release by release drafter. From now on all the changes after 2025.12.0 will be part of all the releases after 2025.12.0.

If we limit release drafter's version sorting algo to sort releases only after GH's latest release (see my commit above), then problem is solved.

OK, if the developer changes calver/semver multiple times after the latest release, during prereleases, then it won't help, but that is quite an unrealistic scenario.

**Comment by cchanche** [comment]
Ok nice, yes I see now how this particular scenario breaks things, and how subsequent releases would append every PRs since v2025.12.0 despite intermediate releases having been published since.

I have been through your code before, and It lead me to suggest the `release-sorting-strategy: release-date` (despite not fully understanding the use-case). I do believe your suggestion is the same as simply looking for the last release in date (=/= _in semver_).

This :

> limit release drafter's version sorting algo to sort releases only after GH's latest release

.. is unnecessary work I believe, but does not break things according to the `release-date` strategy. I left [a small review on your changes](https://github.com/homeassistant-apps/action-release-drafter/commit/c581fd250b8607c7dafece1caa68be1cfae765bd#r174144899), but I may be wrong.

So here is how I see things : 

1. You have a consistant versioning scheme -> use the corresponding strategy (ex: `semver` / `calver`)
2. You have a mixed versioning scheme -> use a generic strategy (ex: `release-date`)

Scenario 1. is more robust in that it will allow you to have different strategies going on at the same time :

- Have the following releases ready (in chonological order) : `app-26.1.0`, `app-26.1.1`, `ui-1.2.3`
- Then trigger a release-drafter run for an app release :
  - with `release-sorting-strategy: "semver"` and `tag-prefix: "app-"`, and ask for a patch
  - publishes `app-26.1.2` with PRs since `app-26.1.1`
  - `app-26.1.1` was correctly identified as the latest release, skipping `ui-1.2.3` for not being calver (to be implemented... but this is for the example)

Scenario 2. won't let you do that : 

- Have the following releases ready (in chonological order) : `app-26.1.0`, `app-26.1.1`, `ui-1.2.3`
- Then trigger a release-drafter run for an app release :
  - with `release-sorting-strategy: "release-date"` (Scenario 2.) and `tag-prefix: "app-"`, and ask for a patch
  - may also publish `app-26.1.2`, but with PRs since `ui-1.2.3` which is not what you want


TLDR : current code finds the latest release with a mix of the `semver` and `release-date` strategies that confuses users on some occasions, and is also not very flexible. I think refactoring into clear strategies, and adding those missing (and forcefully required here) could solve things.

I won't be able to publish anything soon about this, but If this long-a** comment inspires anyone on doing smth, please do 😄

**Comment by AlexSkrypnyk** [comment]
Why do we have to sort releases by version at all? Each release coming from GH API has own date. Can we just sort by that date and always use the latest release to produce the next one?

**Comment by cchanche** [comment]
I agree, that would be much simpler... Had a hard time wrapping my head around it too, but I figured if this was implemented using (at least partly) semver comparison, there must be a reason...

Two tests fail when removing the version-compare part : 

1. `releases` > `findReleases` > `should retrieve last release respecting semver, stripped prefix` [here](https://github.com/release-drafter/release-drafter/blob/b1476f6e6eb133afa41ed8589daba6dc69b4d3f5/test/index.test.js#L1312)
2. `push` > `with include-pre-releases true config` > `includes pre releases` [here](https://github.com/release-drafter/release-drafter/blob/b1476f6e6eb133afa41ed8589daba6dc69b4d3f5/test/releases.test.js#L300)

.. with the following function : 

```js
const sortReleases = (releases, _) => {
  return releases.sort((r1, r2) => {
    return new Date(r1.created_at) - new Date(r2.created_at)
  })
}
```

Test case 1. is about pre-releases and test case 2. is about two previous releases with the same `created_at` date. Pre-releases might be worth investigating, but 2. looks like a coincidence.

It this is to be fixed one way or another, our best bet is most-likely to try make it non-breaking anyways

**Comment by lmagyar** [comment]
> Why do we have to sort releases by version at all? Each release coming from GH API has own date. Can we just sort by that date and always use the latest release to produce the next one?

What if the latest by date is a patch release for a previous version? Eg. v1.0.1, while we are already published v1.5.0 or even v2.0.0?

**Comment by lmagyar** [comment]
> 1. You have a consistant versioning scheme -> use the corresponding strategy (ex: `semver` / `calver`)
> ...
> Scenario 1. is more robust in that it will allow you to have different strategies going on at the same time :
> ...
> `app-26.1.1` was correctly identified as the latest release, skipping `ui-1.2.3` for ***not being calver*** (to be implemented... but this is for the example)

No, skipping it because the tag prefix mismatch. So you don't need a new config to make Scenario 1 work.


> 2. You have a mixed versioning scheme -> use a generic strategy (ex: `release-date`)
> ...
> Scenario 2. won't let you do that :
> ...
> may also publish `app-26.1.2`, but with PRs since `ui-1.2.3` which is not what you want

I don't think so, it just works, because tagPrefix is used for filtering before sorting. So you don't need a new config to make Scenario 2 work. Or depending on implementation of the new `release-sorting-strategy` config, it can even break things.


OK, I clearly don't see something you see, because you really want to introduce a new config and I can't figure out a situation, where it is really needed.

Though my hack has a bug, but I put it in a new comment.

**Comment by lmagyar** [comment]
TLDR

- ***There is a bug in my hack:*** we should drop the GH latest release as a filter, if that has different commitish than the targetCommitish when filterByCommitish is true or has different tag prefix than the tagPrefix when tagPrefix is defined.
- If targetCommitish or tagPrefix is used, my hack (using GH latest release) can't help on the ***other*** "branches", otherwise show me a scenario where it is broken.

---

Problematic scenario: when there is a patch release for a previous release. Previous can be an LTS branch, or just a plain (still supported) previous release.

See this "diagram" stolen from https://github.com/release-drafter/release-drafter/issues/1061#issue-1126478134:
```
    A  B  C            D  F  G  v2.0.0
    |  |  |            |  |  |    |
----o--o--o-----X------o--o--o----X----> default "main" branch
                |   |               
            v1.0.0  |
 (target_commitish   \
     is "main")       \-------o-----------X----> 1.x branch
                              |           |
                              E         v1.0.1 release
```

In this case:
- when v1.0.1 is GH latest, this means LTS/STS is used, this also has to be on a different branch, so commitish should be used
  - my original hack won't see any releases when v2.0.x is released, this is a bug
  - but with the bugfix (see above), it is not activated on the main branch if the GH latest is on a different commitish, so it is still fine
    - though even the fixed hack can't help for calver->semver transition on the STS branch until a new LTS is released (really a corner case)
- if v1.0.1 is not an LTS, this should not be GH latest, just a plain release, v2.0.0 should be GH latest
  - if v1.0.1 is properly released with a different target_commitish, even if it is a calver, it is filtered out by the current rel. drafter code, and even my non-fixed hack works
  - if v1.0.1 has the same commitish (released eg. manually), and is a calver indeed, then repo owner must fix it's commitish to the real branch name to enable my (non-fixed or fixed) hack to properly handle the calver->semver transition

Note: On how to change the target_commitish of an already published release, see: https://github.com/release-drafter/release-drafter/issues/656#issuecomment-806184887

**Comment by cchanche** [comment]
> > Why do we have to sort releases by version at all? Each release coming from GH API has own date. Can we just sort by that date and always use the latest release to produce the next one?
> 
> What if the latest by date is a patch release for a previous version? Eg. v1.0.1, while we are already published v1.5.0 or even v2.0.0?

@lmagyar Please clarify. I think you made quick assumptions here about what release is 
- draft/published
- what their `target_commitish` are
- are they pre-releases

If one is making patches for old major versions, the corresponding release (ex: `v1.0.1`) may appear among the other releases in the release page, but will most-probably not have the same `target_commitish` as `v2.0.0` and `v1.0.0` - thus excluding it (`v1.0.1`) from the resulting releases we have to find the latest from.

  - `v1.0.0`
    - `target_commitish: main`
    - `published: true`
    - `prerelease: false`
  - `v2.0.0`
    - `target_commitish: main`
    - `published: true`
    - `prerelease: false`
  - `v1.0.1`
    - `target_commitish: release/v1.0.0`
    - `published: true`
    - `prerelease: false`
  - `[new draft by release-drafter]`
    - `commitish: main`
    - would find the latest release is `v2.0.0`, whether the sorting strategy is `semver` or `date`
  - `[new draft by release-drafter]`
    - `commitish: release/v1.0.0`
    - would find the latest release is `v1.0.1`, whether the sorting strategy is `semver` or `date`


Here is a nice diagram that showcases what I have in mind (among other things):

<img width="787" height="403" alt="Image" src="https://github.com/user-attachments/assets/4a5b6944-8e86-418e-814c-99e9adb05fa1" />

_source : https://trunkbaseddevelopment.com/branch-for-release/_

**Comment by cchanche** [comment]
> No, skipping it because the tag prefix mismatch. So you don't need a new config to make Scenario 1 work.

Ah, you are correct @lmagyar. I got confused in writing my example ! Introducing `app-` and `ui-` prefixes in my example was meant to showcase that you could have multiple versioning schemes, all while them being consistant (=/= mixed). But since there are prefixes, you do not need to rely on a _versioning-scheme-recognition_ capability to skip incompatible releases from the sorting function - you are correct !

Here is a more accurate version of my scenario 1 showcase ([original comment](https://github.com/release-drafter/release-drafter/issues/913#issuecomment-3725159608)) : 

------

So here is how I see things : 

1. You have a consistant versioning scheme -> use the corresponding strategy (ex: `semver` / `calver`)
2. You have a mixed versioning scheme -> use a generic strategy (ex: `release-date`)

Scenario 1. is more robust in that it will allow you to have different strategies going on at the same time :

- Have the following releases ready (in chonological order) : `v25.12.24` `v26.1.0`, `v26.1.12`, `v1.2.3`
- Then trigger a release-drafter run for a `calver` release :
  - with `release-sorting-strategy: "calver"`
  - publishes/drafts `v26.1.26` (today) - but that is not the interesting part
  - ...what's interesting, is that `v26.1.12` was correctly identified as the latest release, skipping `v1.2.3` for not being calver. That means `v26.1.26` (today) will include changes back up to `v26.1.12` instead of `v1.2.3`

------

If I recall correctly, the original issue was something like : 

- Have the following releases ready (in chonological order) : `v25.12.24` `v26.1.0`, `v26.1.12`, `v1.2.3`

... but `v1.2.3` is meant to be the first `semver` release of a single (mixed) verisoning scheme - instead of what a refered to as "have different strategies going on at the same time". Having a `release-sorting-strategy` would enable doing :

- trigger a release-drafter run for a ~`calver`~ `semver` release :
  - with `release-sorting-strategy: "semver"`
  - publishes/drafts `v1.2.4` (patch bump if so desired) - but that is not the interesting part
  - ...what's interesting, is that `v1.2.3` was correctly identified as the latest release, skipping  all calver releases that would be identified as __greater_/_higher_ by a regular semver-only parser

**Comment by lmagyar** [comment]
We totally agree on 99% of these things. As far as I tested (I'm not a heavy user of rel. drafter), the current implementation works perfectly fine for all the above mentioned scenarios ***if there is no change*** in the strategy.

My hack is a solution when the versioning scheme is ***changing***. And as I realized works only when the GH latest is on the same commitish/tagprefix as the drafter runs on, so sadly not totally general. But at least in that case it is working.

As far as i understand, `release-sorting-strategy` is not needed when there is no change in strategy. So I think this should be an optional argument and used only when the strategy is changed (typically from calver to semver).

**Comment by cchanche** [comment]
> I don't think so, it just works, because tagPrefix is used for filtering before sorting.

Again, sorry @lmagyar for the confusion : I had completely forgotten about `tag-prefixe`s when writing this despite writing examples with literal tag prefixes... my brain is full of surprises !

Here is a more accurate version of what I had in mind for scenario 2 :

1. You have a consistant versioning scheme -> use the corresponding strategy (ex: `semver` / `calver`)
   - both can coexist at the same time on the same `target_commitish`
2. You have a mixed versioning scheme -> use a generic strategy (ex: `release-date`)
   - you mean to have a single scheme, it simply was implemented using multiple _strategies_ over time

[...]

Scenario 2. ~won't let you do that~ : 

- Have the following releases ready (in chonological order) : `v26.1.0`, `v26.1.1`, `v1.2.3`
- Then trigger a release-drafter run :
  - with `release-sorting-strategy: "release-date"`, and ask for a patch (whatever a patch means for you in such a versioning scheme)
  - publishes `v1.2.4`, 
  - but with PRs since `v1.2.3`

**Comment by cchanche** [comment]
> As far as i understand, release-sorting-strategy is not needed when there is no change in strategy. So I think this should be an optional argument and used only when the strategy is changed (typically from calver to semver).

That is exactly how I image it would work, yes - on top of allowing multiple schemes at the same time (without tag-prefixes ! ...which is actually a really weird thing to do now that I think of it...)

edit : this issue is frying my brain right now 🤯 I still need to read [this](https://github.com/release-drafter/release-drafter/issues/913#issuecomment-3765912879) from you @lmagyar

**Comment by lmagyar** [comment]
That comment is the same as your example with the nice drawing. :)

But now I've found a problem with the new `release-sorting-strategy` also. :D This can/should have different values for each commitish/tagprefix "branch", because they can have different strategies. And even changed separately... Pfff. I've already lost to much brain cells on this issue. :(

So back to this comment:

> Can we just sort by that date and always use the latest release to produce the next one?

So doing it after filtering the releases by commitish and tagprefix, shouldn't work? Within a commitish/tagprefix branch, isn't it enough??? I smell problems with pre-releases, I really forgot how it works, maybe latest by date not prerelease (within the already filtered commitish/tagprefix branch)? I'm really lost now. :/

**Comment by cchanche** [comment]
Yeah so basically, the current sorting is `semver` + a fallback to release date.

This (somehow) gets us around the pre-release stuff - which requires additional brain cells to understand as you put-it 😅

What I would love, is : 
- to let people choose between date (only) or semver (only) (if semver is still relevent, ex: prereleases)
- to make clear to pre-release users what to use / what are the implications (if relevant)
- write tests scenarios to make sur we get this right

And then let's just drop this `calver` strategy stuff, this is too much. If one uses such versions, mixed or not inside other schemes, he should be able to use the date sorting like you suggest.

Now the other satellite options, `commitish` and `tag-prefix`, are here to provide tools for users to have multiple schemes in parallel. Let's keep it as-is. No need for this strategy stuff to provide yet another capability towards this.

Scenarios for parallel schemes are :

- The release-branch problem showcased [here](https://github.com/release-drafter/release-drafter/issues/913#issuecomment-3799393744), [here](https://github.com/release-drafter/release-drafter/issues/913#issuecomment-3765912879), [here](https://github.com/release-drafter/release-drafter/issues/1061#issue-1126478134) and [here](https://github.com/release-drafter/release-drafter/pull/1445)
  - same versioning schemes (ex: semver), just multiple instances of it
  - appropriate tool : `target_commitish` (though it does not solve everything)
- the multiple package scenario :
  - ex : `app-26.2.1` and `ui-1.2.3`
  - use prefixes,
- the multiple package scenario, but each package has a different strategy (ex: calver and semver)
  - ex: `26.2.1` and `1.2.3` (two different package releases)
  - migrate to using prefixes, what kind of psycho would think of this ? (me, [above](https://github.com/release-drafter/release-drafter/issues/913#issuecomment-3799458923))


this issue is not about _parallel schemes_, and neither should it's solution. It is about sorting releases in a single / unique scheme, taking into account the following scenarios :

- the scheme has changed it's strategy
  - the current sorting strategy got confused
- prereleases are introduced in the release-tree
  - the pre-release syntax used may not be correct
  - the suggested _GH latest date-only_ strategy got confused
  
am I missing sth ? going forward on this, is this safe for us to consider that every scenario may be in the context of a "ingle / unique scheme" ? I may have forgotten some scenarios in my examples just above, but If we can act that `commitish` and `tag-prefix` are out of scope, that's already something

**Comment by cchanche** [comment]
About the pre-release stuff, here is a scenario I dug out from the tests : 

If you have the following releases (in order) : 

- `v1.5.0-alpha`
  - not released, `prerelease: true`
- `v1.0.0`
  - released, returned by `context.octokit.repos.getLatestRelease`

... what would you want as a next patch version ?

[This test](https://github.com/release-drafter/release-drafter/blob/b1476f6e6eb133afa41ed8589daba6dc69b4d3f5/test/index.test.js#L1312) in the release-drafter showcases that with `include-pre-releases: true`, you'll probably want to publish `v1.5.1`, with changes since `v1.5.0-alpha`.

What this test-case really showcases for us has nothing to do with the `include-pre-releases: true` - this config is only here to justify having `v1.5.0-alpha` **and then** `v1.0.0` in the release-tree. `include-pre-releases: true` only means "this `v1.5.0-alpha`, which is a preprelease, was included in the sorting function"

So now, why was `v1.5.0-alpha` considered as the _latest in date_, despite `v1.0.0` having : 
- a more recent release-date
- being considered as `latest: true` by Github dedicated "latest release" mechanism.

Well, I think this is because of the current sorting function (obviously), that uses semver : 

```js
const sortReleases = (releases) => {
  return releases.sort((r1, r2) => {
    try {
      return compareVersions(r1.tag_name, r2.tag_name)
    } catch {
      return new Date(r1.created_at) - new Date(r2.created_at)
    }
  })
}
```

Docs says that `compareVersions` will return 1 if first version is greater, 0 if versions are equal, and -1 if the second version is greater. In the terminal, I briefly assessed this : 

```bash
Welcome to Node.js v20.20.0.
Type ".help" for more information.
> const compareVersions = require('compare-versions')
undefined
> compareVersions('v1.5.0-alpha', 'v1.0.0')
1
```


... which explains how the current sorting function simply _works_ for pre-releases.


Now, with this in mind, we need to find something to solve the following cases. `input` is whatever enters the sorting function. It may have been altered beforehand (ex: adding pre-releases, exluding some prefixes, comitish, etc...). For the sake of this table's clarity, I have represented this `input` column as though they would have been sorted by release-date already. This is obvisouly not the case, but helps understanding the `res` column whenever `strategy` is `date`. This strategy is not in the code, but was included here as a proposal to try and find solutions

| #   | input                               | strategy | res               | ok ? |
| --- | ----------------------------------- | -------- | ----------------- | ---- |
| 1   | `['v1.0.0', 'v1.5.0']`              | current  | `'v1.5.0'`        | yes  |
| 2   | `['v1.5.0', 'v1.0.0']`              | current  | `'v1.5.0'`        | yes  |
| 3   | `['v1.5.0-alpha', 'v1.0.0']`        | current  | `'v1.5.0-alpha'`  | yes  |
| 4   | `['v1.0.0', 'v1.5.0-alpha']`        | current  | `'v1.5.0-alpha'`  | yes  |
| 5   | `['v2022.04.28', 'v1.0.0']`         | current  | `'v2022.04.28'`   | no - but expected  |
| 6   | `['v2022.04.28', 'v1.0.0']`         | date     | `'v1.0.0'`        | yes  |
| 7   | `['3.0.0-beta-9', '3.0.0-beta-14']` | current  | `'3.0.0-beta-9'`  | ?    |
| 8   | `['3.0.0-beta-9', '3.0.0-beta-14']` | date     | `'3.0.0-beta-14'` | yes  |
| 9   | `['3.0.0-beta-1', '3.0.0-beta-14']` | current  | `'3.0.0-beta-14'` | ?    |
| 10  | `['3.0.0-beta-1', '3.0.0-beta-14']` | current  | `'3.0.0-beta-14'` | ?    |
| 11  | `['3.0.0-beta.1', '3.0.0-beta.14']` | current  | `'3.0.0-beta.14'` | ?    |
| 12  | `['3.0.0-beta.14', '3.0.0-beta.1']` | current  | `'3.0.0-beta.14'` | ?    |

**Comment by cchanche** [comment]
@oleg-nenashev I fear `3.0.0-beta-14` may not be the syntax you need for your pre-releases, and that it may be causing you trouble.

From the table above (and my tests with compare-version), I think you may want to use `3.0.0-beta.14` instead. (note the `.` replacing your last `-`)

Here are some more examples : 

```bash
> compareVersions('3.0.0-beta-123', '3.0.0-beta-12' )
1 
> compareVersions('3.0.0-beta-123', '3.0.0-beta-13' )
-1
```

...but : 


```bash
> compareVersions('3.0.0-beta.123', '3.0.0-beta.12' )
1
> compareVersions('3.0.0-beta.123', '3.0.0-beta.13' )
1
```


It looks to me as if : 
- `*-beta-[n]` expects n to be a semver,  ex: `1.2.3` or `1.2`, but ignores the missing dot separators
- `*-beta.[n]` expects n to be a simple decimal number counting upwards,  ex: `1` < `12` < `13` < `123`

**Comment by lmagyar** [comment]
Note:
```
  const sortedSelectedReleases = sortReleases(
    filteredReleases.filter(
      (r) => !r.draft && (!r.prerelease || includePreReleases)
```
So the config decides whether we "see" prereleases or not.

***Anything that has later date than what we "want", should be on another commitish or tagprefix, or be filtered out by prerelease.***

So, isn't it enough?
``` diff
-const sortReleases = (releases, tagPrefix) => {
+const sortReleases = (releases) => {
-  // For semver, we find the greatest release number
-  // For non-semver, we use the most recently merged
-  const tagPrefixRexExp = new RegExp(`^${regexEscape(tagPrefix)}`)
  return releases.sort((r1, r2) => {
-    try {
-      return compareVersions(
-        r1.tag_name.replace(tagPrefixRexExp, ''),
-        r2.tag_name.replace(tagPrefixRexExp, '')
-      )
-    } catch {
      return new Date(r1.created_at) - new Date(r2.created_at)
-    }
  })
}
```

Or just drop the whole sortRelease stuff and go with (not tested, just edited here):
```
const sortedSelectedReleases = 
  filteredReleases.filter(
    (r) => !r.draft && (!r.prerelease || includePreReleases)
  ).sort(
    (r1, r2) => new Date(r1.created_at) - new Date(r2.created_at)
  )
```

And when people want to change from calver to semver, they just edit the release title/tag, and from now on that is used.

Nothing to configure, just works, maybe some documentation needed, that in case a wrong version is selected, think about using commitish or tagprefix properly.

**Comment by cchanche** [comment]
> **Anything that has later date than what we "want", should be on another commitish or tagprefix, or be filtered out by prerelease.**

I am really unsure I understand that, I am sorry... As stated [here](https://github.com/release-drafter/release-drafter/issues/913#issuecomment-3799825717), I really think the `commitish`, `tagprefix` and `includePreReleases` are out of scope for this issue.

Both of your snippets suggest we implement _date-only_ sorting for releases. Assuming this is what we went for, here is how every test-case (except number 6 & 8) would go : 


| #   | input                                 | strategy | res                 | ok ?              |
| --- | ------------------------------------- | -------- | ------------------- | ----------------- |
| 1   | `['v1.0.0', 'v1.5.0']`                | current  | `'v1.5.0'`          | yes               |
| 2   | `['v1.5.0', 'v1.0.0']`                | current  | `'v1.5.0'`          | yes               |
| 3   | `['v1.5.0-alpha', 'v1.0.0']`          | current  | `'v1.5.0-alpha'`    | yes               |
| 4   | `['v1.0.0', 'v1.5.0-alpha']`          | current  | `'v1.5.0-alpha'`    | yes               |
| 5   | `['v2022.04.28', 'v1.0.0']`           | current  | `'v2022.04.28'`     | no - but expected |
| 6   | `['v2022.04.28', 'v1.0.0']`           | date     | `'v1.0.0'`          | yes               |
| 7   | `['3.0.0-beta-9', '3.0.0-beta-14']`   | current  | `'3.0.0-beta-9'`    | ?                 |
| 8   | `['3.0.0-beta-9', '3.0.0-beta-14']`   | date     | `'3.0.0-beta-14'`   | yes               |
| 9   | `['3.0.0-beta-1', '3.0.0-beta-14']`   | current  | `'3.0.0-beta-14'`   | ?                 |
| 10  | `['3.0.0-beta-1', '3.0.0-beta-14']`   | current  | `'3.0.0-beta-14'`   | ?                 |
| 11  | `['3.0.0-beta.1', '3.0.0-beta.14']`   | current  | `'3.0.0-beta.14'`   | ?                 |
| 12  | `['3.0.0-beta.14', '3.0.0-beta.1']`   | current  | `'3.0.0-beta.14'`   | ?                 |
| --- | -----------------------------------   | -------- | -----------------   | ----              |
| 13  | `['v1.0.0', 'v1.5.0']`                | date     | `'v1.5.0'`          | yes               |
| 14  | `['v1.5.0', 'v1.0.0']`                | date     | `'v1.0.0'`          | no                |
| 15  | `['v1.5.0-alpha', 'v1.0.0']`          | date     | `'v1.0.0'`          | no                |
| 16  | `['v1.0.0', 'v1.5.0-alpha']`          | date     | `'v1.5.0-alpha'`    | yes               |
| 17  | `['v2022.04.28', 'v1.0.0']`           | date     | `'v1.0.0'`          | yes               |
| ~6~ | ~`['v2022.04.28', 'v1.0.0']`~         | ~date~   | ~`'v1.0.0'`~        | ~yes~             |
| 18  | `['3.0.0-beta-9', '3.0.0-beta-14']`   | date     | `'3.0.0-beta-14'`   | ?                 |
| ~8~ | ~`['3.0.0-beta-9', '3.0.0-beta-14']`~ | ~date~   | ~`'3.0.0-beta-14'`~ | ~yes~             |
| 20  | `['3.0.0-beta-1', '3.0.0-beta-14']`   | date     | `'3.0.0-beta-14'`   | ?                 |
| 21  | `['3.0.0-beta-1', '3.0.0-beta-14']`   | date     | `'3.0.0-beta-14'`   | ?                 |
| 22  | `['3.0.0-beta.1', '3.0.0-beta.14']`   | date     | `'3.0.0-beta.14'`   | ?                 |
| 23  | `['3.0.0-beta.14', '3.0.0-beta.1']`   | date     | `'3.0.0-beta.1'`    | ?                 |


you can see test case 14 & 15 fail

**Comment by lmagyar** [comment]
> > **Anything that has later date than what we "want", should be on another commitish or tagprefix, or be filtered out by prerelease.**
> 
> I am really unsure I understand that, I am sorry... As stated [here](https://github.com/release-drafter/release-drafter/issues/913#issuecomment-3799825717), I really think the `commitish`, `tagprefix` and `includePreReleases` are out of scope for this issue.

I think they are essential to this issue! :D They narrow the whole release "space" to a subset (`filteredReleases`), that contains the releases that are relevant calculating the last/next release. Thankfully tagprefix creates distinct subsets, while commitish and prerelease handles "branching" subsets. "Branching" like the patch release, that has later date, but won't be on the main branch.

> | # | input | strategy | res | ok ? |
> |-|-|-|-|-|
> | 14 | `['v1.5.0', 'v1.0.0']` | date | `'v1.0.0'` | no |

Shouldn't v1.0.0 be on a different commitish as a patch release? So shouldn't be part of the sorting subset?

Hmmmm, or it is a tag on the same git branch, just it happened to be released later??? Can this happen? Even if the v1.0.0 tag date is smaller than v1.5.0, the release date will be later? I don't know, just asking.

> | # | input | strategy | res | ok ? |
> |-|-|-|-|-|
> | 15 | `['v1.5.0-alpha', 'v1.0.0']` | date | `'v1.0.0'` | no |

Same as the previous.

Maybe more exciting timeline:
- v1.0.0
- v1.5.0-alpha
- v1.1.0 we decided to make a real release before v1.5.0

But v1.1.0 can't be on the same git branch. Maybe we tag v1.1.0 on the same linear git branch and release it after v1.5.0-alpha??? This seems to be realistic, but quite ugly.

And we continue the timeline:
- v2.0.0 because v1.5.0 never got released

This can be, should be on the same git branch as v1.5.0 was.

TLDR:

I have the feeling, that:
- if commitish is not defined, we go with the current sorting, backward compatible, because even if the releases we see in filteredReleases are on different git branches, they have the same commitish, because nobody cared to set it properly
- if commitish is used, we can assume users know what they are doing and we go with the date-only sorting, this helps even if there is only 1 linear git branch, because solves eg. the calver->semver transition

**Comment by cchanche** [comment]
> They narrow the whole release "space" to a subset (filteredReleases)

exactly ! Our issue is with what happens next : the sorting.

> Shouldn't v1.0.0 be on a different commitish as a patch release

As weird as it may appear (as you demonstrated) it is not. (at least in the current tests) Please see [this test suite](https://github.com/release-drafter/release-drafter/blob/40bb31b945af999c14bc1339eae4c0371d97cac7/test/index.test.js#L1281). Both have `"target_commitish": "master"`

This may indeed be a rather unrealistic scenario, but I think it may not matter that much. The point of this test suite is to demonstrate `compareVersion()` and the corresponding sorting function does it's job (as implemented).

Now, maybe this implementation needs to change, but again : it would be better if it is non-breaking. There may be a real-world scenario I (or we) do not see [here](https://github.com/release-drafter/release-drafter/blob/40bb31b945af999c14bc1339eae4c0371d97cac7/test/index.test.js#L1281).

**Comment by cchanche** [comment]
> | # | input | strategy | res | ok ? |
> |-|-|-|-|-|
> | 14 | `['v1.5.0', 'v1.0.0']` | date | `'v1.0.0'` | no |
> | 15 | `['v1.5.0-alpha', 'v1.0.0']` | date | `'v1.0.0'` | no |

Just to be clear, `ok? no` here means : if the new sorting strategy becomes `date` **as the only strategy available**, tests fail.

What I was proposing is to have the option of choosing the strategy accoring to one's needs, and keep the dfault one as-is. Scenarios 14 & 15 may be `ok? yes` for some of us (if they decide the `date` strategy).

**Comment by lmagyar** [comment]
I'm back and I agree.

And you don't have to wait for me to move on with the new strategy option, I'm really just newbie user of rel. drafter. My hack (using GH latest release) had proven broken in case of properly using commitish and/or tagprefix, my idea to automatically turn on date sorting can potentially break existing configs.

Maybe the main conclusion can be, that only 2 values are needed for the new strategy option: version and date (where version is the current algo).

**Comment by jetersen** [comment]
I think this should be mostly resolved by #1570 if that is not the case, I would be surprised.

**Comment by lmagyar** [comment]
> I think this should be mostly resolved by [#1570](https://github.com/release-drafter/release-drafter/pull/1570) if that is not the case, I would be surprised.

Sadly you should be surprised, this is still totally broken!!!

This issue has to be reopened.

See: https://github.com/homeassistant-apps/app-newt/releases

After v1.0.8 (2 weeks ago) it creates a draft release as v2025.11.4, because half year ago v2025.11.3 was the last semver version.

And it collects all the PRs from the last year, not only from the last 2 weeks.

This is as broken az is was previously.

**Comment by lmagyar** [comment]
Just for reference what is the current draft release (because you won't see the draft).

Title: v2025.11.4

Body:

- ⬆️ Update newt to v1.11.0 @[renovate[bot]](https://github.com/apps/renovate) ([#54](https://github.com/homeassistant-apps/app-newt/pull/54))

### 🧰 Maintenance


- Update CI to org repo @elcajon ([#33](https://github.com/homeassistant-apps/app-newt/pull/33))
- Modify release drafter template @lmagyar ([#36](https://github.com/homeassistant-apps/app-newt/pull/36))
- Run release drafter as last job in CI @lmagyar ([#38](https://github.com/homeassistant-apps/app-newt/pull/38))
- Fix copy paste in translation @brenner-tobias ([#41](https://github.com/homeassistant-apps/app-newt/pull/41))
- Remove legacy architectures @brenner-tobias ([#42](https://github.com/homeassistant-apps/app-newt/pull/42))
- Add PR's url to the release draft @lmagyar ([#43](https://github.com/homeassistant-apps/app-newt/pull/43))
- Use homeassistant-apps/.github repo's release-drafter.yml @lmagyar ([#49](https://github.com/homeassistant-apps/app-newt/pull/49))
- Use official release drafter, as a standalone workflow, without delay @lmagyar ([#70](https://github.com/homeassistant-apps/app-newt/pull/70))

### 📚 Documentation


- [WIP] Address feedback on copyright update in PR #34 @[copilot-swe-agent[bot]](https://github.com/apps/copilot-swe-agent) ([#35](https://github.com/homeassistant-apps/app-newt/pull/35))
- Update copyright @elcajon ([#34](https://github.com/homeassistant-apps/app-newt/pull/34))
- Fix copy paste in translation @brenner-tobias ([#41](https://github.com/homeassistant-apps/app-newt/pull/41))

### ⬆️ Dependency updates


- ⬆️ Update newt to v1.7.0 @[renovate[bot]](https://github.com/apps/renovate) ([#37](https://github.com/homeassistant-apps/app-newt/pull/37))
- ⬆️ Update newt to v1.8.0 @[renovate[bot]](https://github.com/apps/renovate) ([#39](https://github.com/homeassistant-apps/app-newt/pull/39))
- ⬆️ Update newt to v1.8.1 @[renovate[bot]](https://github.com/apps/renovate) ([#40](https://github.com/homeassistant-apps/app-newt/pull/40))
- ⬆️ Update newt to v1.9.0 @[renovate[bot]](https://github.com/apps/renovate) ([#44](https://github.com/homeassistant-apps/app-newt/pull/44))
- ⬆️ Update ghcr.io/hassio-addons/base Docker tag to v20 @[renovate[bot]](https://github.com/apps/renovate) ([#45](https://github.com/homeassistant-apps/app-newt/pull/45))
- ⬆️ Update ghcr.io/hassio-addons/base Docker tag to v20.0.1 @[renovate[bot]](https://github.com/apps/renovate) ([#46](https://github.com/homeassistant-apps/app-newt/pull/46))
- ⬆️ Update newt to v1.10.0 @[renovate[bot]](https://github.com/apps/renovate) ([#47](https://github.com/homeassistant-apps/app-newt/pull/47))
- ⬆️ Update newt to v1.10.1 @[renovate[bot]](https://github.com/apps/renovate) ([#48](https://github.com/homeassistant-apps/app-newt/pull/48))
- ⬆️ Update newt to v1.10.2 @[renovate[bot]](https://github.com/apps/renovate) ([#50](https://github.com/homeassistant-apps/app-newt/pull/50))
- ⬆️ Update newt to v1.10.3 @[renovate[bot]](https://github.com/apps/renovate) ([#51](https://github.com/homeassistant-apps/app-newt/pull/51))
- ⬆️ Update newt to v1.10.4 @[renovate[bot]](https://github.com/apps/renovate) ([#52](https://github.com/homeassistant-apps/app-newt/pull/52))
- ⬆️ Update ghcr.io/hassio-addons/base Docker tag to v20.0.2 @[renovate[bot]](https://github.com/apps/renovate) ([#53](https://github.com/homeassistant-apps/app-newt/pull/53))
- ⬆️ Update ghcr.io/hassio-addons/base Docker tag to v20.0.3 @[renovate[bot]](https://github.com/apps/renovate) ([#55](https://github.com/homeassistant-apps/app-newt/pull/55))
- ⬆️ Update ghcr.io/hassio-addons/base Docker tag to v20.0.4 @[renovate[bot]](https://github.com/apps/renovate) ([#56](https://github.com/homeassistant-apps/app-newt/pull/56))
- ⬆️ Update ghcr.io/hassio-addons/base Docker tag to v20.1.0 @[renovate[bot]](https://github.com/apps/renovate) ([#57](https://github.com/homeassistant-apps/app-newt/pull/57))
- ⬆️ Update newt to v1.12.0 @[renovate[bot]](https://github.com/apps/renovate) ([#58](https://github.com/homeassistant-apps/app-newt/pull/58))
- ⬆️ Update newt to v1.12.1 @[renovate[bot]](https://github.com/apps/renovate) ([#59](https://github.com/homeassistant-apps/app-newt/pull/59))
- ⬆️ Update newt to v1.12.2 @[renovate[bot]](https://github.com/apps/renovate) ([#60](https://github.com/homeassistant-apps/app-newt/pull/60))
- ⬆️ Update newt to v1.12.3 @[renovate[bot]](https://github.com/apps/renovate) ([#61](https://github.com/homeassistant-apps/app-newt/pull/61))
- ⬆️ Update ghcr.io/hassio-addons/base Docker tag to v20.1.1 @[renovate[bot]](https://github.com/apps/renovate) ([#62](https://github.com/homeassistant-apps/app-newt/pull/62))
- ⬆️ Update newt to v1.12.4 @[renovate[bot]](https://github.com/apps/renovate) ([#63](https://github.com/homeassistant-apps/app-newt/pull/63))
- ⬆️ Update newt to v1.12.5 @[renovate[bot]](https://github.com/apps/renovate) ([#64](https://github.com/homeassistant-apps/app-newt/pull/64))
- ⬆️ Update Add-on base image to v20.2.0 @[renovate[bot]](https://github.com/apps/renovate) ([#65](https://github.com/homeassistant-apps/app-newt/pull/65))
- ⬆️ Update newt to v1.13.0 @[renovate[bot]](https://github.com/apps/renovate) ([#67](https://github.com/homeassistant-apps/app-newt/pull/67))
- ⬆️ Update Add-on base image to v21 @[renovate[bot]](https://github.com/apps/renovate) ([#66](https://github.com/homeassistant-apps/app-newt/pull/66))
- ⬆️ Update newt to v1.14.0 @[renovate[bot]](https://github.com/apps/renovate) ([#68](https://github.com/homeassistant-apps/app-newt/pull/68))
- ⬆️ Update newt to v1.15.0 @[renovate[bot]](https://github.com/apps/renovate) ([#69](https://github.com/homeassistant-apps/app-newt/pull/69))

**Comment by jetersen** [comment]
You changed the version schema from cal version to v1 so obviously it's on you 😅
I'd suggest adding `filter-by-range` to your workflow action input or config parameter.

**Comment by cchanche** [comment]
Yeah filter-by-range would help, but I recall the original direction our discussion went in was to implement different versioning strategies to accommodate `semver` (default), `date` or `calver`, etc... This would allow filtering-out versions that do not mach the configured strategy.

I am unsure what the actual conclusion was, but I kept in mind that this was technically too hard, and that you should instead rely on external versioning actions, and input your version using the `version:` input directly.

**Comment by jetersen** [comment]
well in this case they clearly switch from v2025 to v1.
Two options exist, cleanup your old releases that use v2025 or use a semver range.

**Comment by AlexSkrypnyk** [comment]
@jetersen 
> Two options exist, cleanup your old releases that use v2025 or use a semver range.

I'm not sure I'm following what you are suggesting.
This whole issue was about resolving a next version **after a switch between versions schemes took place**. Suggesting "cleanup your old releases" is not a solution (otherwise we would not be discussing this issue).

What is the real solution here? On Jan 6, I [provided](https://github.com/release-drafter/release-drafter/issues/913#issuecomment-3713948842) 4 options to resolve this. Has any of them been considered by the maintainers?

**Comment by jetersen** [comment]
@AlexSkrypnyk semver range selector is fully documented: `filter-by-range` is something you can drop into your release-drafter.yml or github workflow action input.

As an example for homeassistant-apps/app-newt:

Add this to .github/release-drafter.yml:

```yaml
filter-by-range: "<2000.0.0"
```

That is already present in homeassistant-apps/.github’s shared configuration and is a good choice:

- v1.0.8 is coerced to 1.0.8 and included.
- v2025.11.3 is coerced to 2025.11.3 and excluded.
- Future conventional versions such as v2.0.0 remain eligible.

Release Drafter otherwise sorts by SemVer, so v2025.11.3 would outrank v1.0.8 despite being older. The filtering occurs before [selecting the greatest published version](https://github.com/release-drafter/release-drafter/blob/767bd2518af5a113181a0cbc8e07bf8beabcbcd7/src/actions/drafter/lib/find-previous-releases/find-previous-releases.ts#L77-L103).

If releases must remain strictly within major version 1, use:

```yaml
filter-by-range: "1.x"
```

But that would need changing when moving to v2, so <2000.0.0 is preferable for the CalVer-to-SemVer transition.

Multiple ranges are also supported:
```yaml
filter-by-range: ">=1.0.0 <2000.0.0"
```
